### DISCUSSION TITLE
setQueryData can create "dead" queries

### DISCUSSION THREAD

**Discussion by phatmann** [TanStack/query#2823] [General]
I have two different queries in my app. One query gets messages and one gets conversations. 

```javascript
useQuery('messages', fetchMessages)
useQuery('conversations', fetchConversations)
```

Each conversation object needs to get populated with the last received message, if any. After the messages are fetched, I look through the conversations cache and then update it appropriately. 

```javascript
useQuery('messages', fetchMessages, {
  onSuccess: (messages) => setQueryData('conversations', existingConversations => {
    if (!existingConversations) {
      return
    }
    
    return updateWithRecentMessages(existingConversations, messages)
})
```

When the app started up, the messages query would fire before the first conversations query, so `existingConversations` would be undefined. This undefined value then gets written to the conversations cache. For some reason, this then causes the query function for 'conversations' to never fire. That is, React Query never marks the undefined conversation cache as stale.

Once I realized this, the workaround was easy: don't use an updater function. So the correct code is:

```javascript
useQuery('messages', fetchMessages, {
  onSuccess: (messages) => {
    const existingConversations = client.getQueryData('conversations')

    if (!existingConversations) {
      return
    }

    setQueryData('conversations', updateWithRecentMessages(existingConversations, messages))
  }
}
```

I would like other users to avoid this issue, since it caused me a lot of pain and head-scratching. One idea: if the updater function returns `undefined`, do not set the cache data. If this seems like the right approach, I am happy to make a PR for it.

The combination of the updater function and undefined caches has come up many times. Here are the related discussions and issues:

* https://github.com/tannerlinsley/react-query/discussions/925
* https://github.com/tannerlinsley/react-query/discussions/834
* https://github.com/tannerlinsley/react-query/issues/486
* https://github.com/tannerlinsley/react-query/issues/506
* https://github.com/tannerlinsley/react-query/issues/453
* https://github.com/tannerlinsley/react-query/issues/109

I am starting to wonder if having an updater function syntax is a good idea or not. It was created to be "convenient", but the issues with Typescript and undefined caches make it seem less than ideal.

**Comment by TkDodo** [comment]
> One idea: if the updater function returns undefined, do not set the cache data.

As you said, we've been there before :) One of the issues you've linked was also opened by me a long time ago :)
One problem is that `undefined` could be a legit value for a cached entry 🤷‍♂️. We try not to take any assumptions on that (I think).

The current implementation is made that if you call `setQueryData`, the cache entry will be created. There is no "bailing out" of it.

> Once I realized this, the workaround was easy: don't use an updater function. So the correct code is:

That is also my approach, because with TypeScript, it's very hard to work with the `undefined` stuff. Maybe if we want bailing out, what we could do is something that react useState is also doing: If you provide the referentially same value, we don't update. This would work for `undefined`, but also, if the current value is `5`, and you set it to `5` as well, we would bail.

The problem could be in the details: looking at the code, we build a cache entry first, and then call setData on it:

https://github.com/tannerlinsley/react-query/blob/255f7c66a8d4a5c33a3b167304f787c7c4acf6f5/src/core/queryClient.ts#L136-L139

since `setData` needs access to the previous value. So I think it's possible, but we would have to shuffle the code around a bit :)

**Reply by phatmann** [reply]
Thanks for all the detailed thinking. 

> One problem is that undefined could be a legit value for a cached entry 🤷‍♂️. We try not to take any assumptions on that (I think).

There seems to be an assumption that the cached value will never be undefined:

- To see if a cache exists, you call `getQueryData` and see if it is undefined. 
- When I inadvertently set the cached value to undefined, future queries in that cache did not go stale and call the query function.

Since `setQueryData` always creates a cache, setting the cached value to undefined could be treated as an error. If we do this, and the caller of `setQueryData` cannot handle the case of a nonexistent cache in the updater function, then they will need to check with `getQueryData` first. 

On the other hand, we could still allow `null` cache values.

**Reply by TkDodo** [reply]
> When I inadvertently set the cached value to undefined, future queries in that cache did not go stale and call the query function.

can you create a quick codesandbox example that shows this behaviour please?

> Since setQueryData always creates a cache, setting the cached value to undefined could be treated as an error. 

So, what's your proposal to actually do if returning `undefined` from the `setQueryData` function? With the current behaviour, the cache entry has already been created. I think we'd either way need to inverse the behaviour somehow:
1. get the current value
1. call the updater with the current value
1. check for a legit value
1. create / update the cache entry if the value is legit and inform observers about it.

As far as I understand, your proposal would be to check that a value is legit if it's something else than `undefined`. I would go further and say that we don't need to do anything (create or update the cache) if the value is equal (Object.is) with the previous value. This seems to be in line with what useState is doing, see: https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-state-update
And it would also work if we have undefined currently (no entry exists) and return undefined.

**Reply by TkDodo** [reply]
But I have to concur that we do take some assumptions, given that if you return `undefined` from `initialData`, the cache entry is _not_ created, while the same doesn't work for `setQueryData`.

However, `undefined` seems to be a valid cache value. See this simple codesandbox fork, where I return `Promise.resolve(undefined)`: https://codesandbox.io/s/keen-colden-sngus?file=/src/index.js

**Reply by TkDodo** [reply]
One thing that comes to mind that might be troublesome with the Obj.is comparison is the following scenario:

- you have a cached query with value `5`
- you call `setQueryData<number>` on it with the updater, the signature will be: `(prev: number | undefined) => number | undefined`
- you can now legitimately return `undefined`, and that will be written to the cache, even though you only want numbers....

I really don't see a way around this though, except always bailing out if _undefined_ was returned, which would make _undefined_ be an "illegal" value for cache entries

**Reply by phatmann** [reply]
Making `undefined` illegal seems quite acceptable. In fact I already assumed this was the case before this thread started.

**Reply by TkDodo** [reply]
okay, let's settle for that then. I think having a query in `success` state with `undefined` as `data` is kind of an invariant that shouldn't happen. So, to summarize:

- let's allow the `setQueryData` functional updater to return `undefined`
  - signature: `T | ((prev: T | undefined) => T | undefined)`
  - if `undefined` is returned, we _do not_ create or update the cache entry
  - that means if we have a previous value and `undefined` is returned - `undefined` is _not_ written to the cache, and observers are not informed about that.
  - this is technically breaking if `undefined` is a legit cache entry value for someone, but I'd accept that. I still think it's a "fix".
  - as a further optimization, let's also "bail out" if we're equal with `Object.is` (react has a good [implementation including a polyfill](https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/shared/objectIs.js)).

would you like to contribute that @phatmann ?

**Reply by phatmann** [reply]
Sounds perfect. I'll put a PR together.

**Reply by phatmann** [reply]
@TkDodo There is already code to compare objects when setting cache data:

```typescript {.line-numbers}
    // Use prev data if an isDataEqual function is defined and returns `true`
    if (this.options.isDataEqual?.(prevData, data)) {
      data = prevData as TData
    } else if (this.options.structuralSharing !== false) {
      // Structurally share data between prev and new data if needed
      data = replaceEqualDeep(prevData, data)
    }
```

In order to bail out when data is equal, I would bail out when `isDataEqual` is present and returns true, or when `replaceEqualDeep` returns prevData (which it does if they are deeply equal). Do you concur?

Would it be cleaner if I did two different PRs, one for "undefined" bail out and one for "data equal" bail out? Or do they go together in your mind?

**Reply by phatmann** [reply]
If `isDataEqual` is not provided and `structuralSharing` is false, should we use `Object.is` to see if we should bail out? Or does no structural sharing imply that no comparison should be performed?

**Reply by phatmann** [reply]
I am almost done with the PR. However, there is an item I want to discuss. `setQueryData` uses `query.setData` to update the cached data. In order to provide the "bail out" feature, I made the changes in query.setData because that is where the relevant code is. After a query function is called, `query.setData` is also used to update cache. This means that a query function that returns either `undefined` or the previously cached value will not update the cache and will not cause `onSuccess` or `onSettled` to be called. Since this is breaking behavior, I want to make sure this is acceptable. @TkDodo @tannerlinsley

**Reply by TkDodo** [reply]
@phatmann thanks for looking into this.

I would imagine that we just alway compare with Object.is after structural sharing, something like:

```
    // Use prev data if an isDataEqual function is defined and returns `true`
    if (this.options.isDataEqual?.(prevData, data)) {
      data = prevData as TData
    } else if (this.options.structuralSharing !== false) {
      // Structurally share data between prev and new data if needed
      data = replaceEqualDeep(prevData, data)
    }

if (Object.is(prevData, data)) { set it... }
```

> I am almost done with the PR. However, there is an item I want to discuss. setQueryData uses query.setData to update the cached data. In order to provide the "bail out" feature, I made the changes in query.setData because that is where the relevant code is. After a query function is called, query.setData is also used to update cache. This means that a query function that returns either undefined or the previously cached value will not update the cache and will not cause onSuccess or onSettled to be called. Since this is breaking behavior, I want to make sure this is acceptable.

We're currently planning v4 and I think this feature / breaking change is on the list. I think we might need to wait for v4 instead of merging this PR directly to v3.

**Reply by phatmann** [reply]
Sounds like a good plan. I will finish the PR so you guys can take a look.

**Reply by TkDodo** [reply]
thank you. I just thought about `disabled` queries: If you have a disabled query, it will create a cache entry in `idle` state with `undefined` data. But I don't think it's a problem, right? Currently, if you return `undefined` from `setQueryData`, the query will go into `success` state, and with the new behaviour, it would bail out and stay in `idle` state, which I think is more correct ... right?

**Reply by phatmann** [reply]
I cannot see any inherent problem with an idle cache that has undefined data. A user calling `getQueryData` might think cache does not exist, but if they really care, they can use `QueryCache.find`.

**Reply by phatmann** [reply]
When writing unit tests, I noticed that calling `setQueryData` does not call the `onSuccess` callback for the cache. (As discussed, the `onSuccess` callback is called for each query observer). The documentation says that the cache's `onSuccess` function "will be called if some query is successful". So I am not sure if this is a bug or not. I think for consistency the cache's `onSuccess` callback should be called for every `setQueryData` call that triggers the observers' `onSuccess`. This is what I will do in the PR. And I will update the docs as well to match observer `onSuccess` criteria.

**Reply by phatmann** [reply]
The PR is ready to review! Check it out: https://github.com/tannerlinsley/react-query/pull/2905 @tannerlinsley @TkDodo

**Reply by TkDodo** [reply]
@phatmann yes, that sounds like a bug. I think I just forgot about that when I introduced that callback, because there is no such thing for the onError callback 😅
